Update `gleam_json` to 2.x

SnakeDoc opened the pull request:
This PR closes #4 

This PR is broken down into a few steps:

1. Updates all in-range deps.
2. Updates `gleam_json`'s range to `2.x` along with it's locked version.
3. Fixes a [deprecation introduced by `gleam_stdlib`](https://hexdocs.pm/gleam_stdlib/gleam/result.html#nil_error) update.
4. Updates the GitHub Action's testing environment (gleam and erlang versions).
5. Bumps semvar major version of this lib due to the [new OTP27 requirement](https://hexdocs.pm/gleam_json/).

- [x] Builds Locally
- [x] Tests Run Locally

Let me know if I missed anything or you need something changed! :rocket:

benjaminjellis commented:
It would be really useful to get this merged!

SnakeDoc commented:
@brettkolodny Have you had a chance to review this PR yet? Any changes you need?

brettkolodny commented:
Sorry! Been a bit MIA since the holidays. Going to take a look now!

brettkolodny commented:
@SnakeDoc  All looks good! One thing though, should the `gleam_json` version be instead set to `gleam_json >= 1.0.0 and < 3.0.0`? I believe the only breaking change for `gleam_json` 1.0 to 2.0 is moving what it's using on the Erlang side. May be good to be compatible with older projects!

SnakeDoc commented:
@brettkolodny Great suggestion! I think I was just copying the existing range pattern, but you are right, it's best to be compatible with the entire 2.x range.

Let me get that fixed up sometime tonight. I'll also re-update everything else while we have the hood open :rocket:

SnakeDoc commented:
@brettkolodny Added the following:

1. Updated all deps.
2. Widened the `gleam_json` range as suggested.
3. Took the liberty of converting all dep version ranges to compound ranges. Compound ranges are a bit easier widen over time, and more closely matches what the gleam cli uses.
4. Fixed deprecations introduced by `gleam_stdlib` updates (again lol). This time we made the [`gleam/dynamic` -> `gleam/dynamic/decode` switch](https://hexdocs.pm/gleam_stdlib/gleam/dynamic.html#field), and replaced [`list.pop()` with `list.first() + list.drop()`](https://hexdocs.pm/gleam_stdlib/gleam/list.html#pop).
5. Updated GH Action's erlang and gleam versions.

Builds with all tests passing locally - but this has a bit more surface area than before and probably warrants a closer look.

Let me know what you think!

brettkolodny commented:
Looks good to me, thank you! Will get this released now

SnakeDoc commented:
Thank you 👍
